pkg/kubelet/logs/container_log_manager.go

```diff
@@ -0,0 +1,401 @@
+	// Sort log files in oldest to newest order.
+	sort.Strings(logs)
+	// Container will create a new log file, and we'll rotate the latest log file.
+	// Other than those 2 files, we can have at most MaxFiles-2 rotated log files.
+	// Keep MaxFiles-2 files by removing old files.
+	// We should remove from oldest to newest, so as not to break ongoing `kubectl logs`.
+	maxRotatedFiles := c.policy.MaxFiles - 2
+	if maxRotatedFiles < 0 {
+		maxRotatedFiles = 0
+	}
+	i := 0
+	for ; i < len(logs)-maxRotatedFiles; i++ {
+		if err := os.Remove(logs[i]); err != nil {
+			return nil, fmt.Errorf("failed to remove old log %q: %v", logs[i], err)
+		}
+	}
+	logs = logs[i:]
+	return logs, nil
+}
+
+// compressLog compresses a log to log.gz with gzip.
+func (c *containerLogManager) compressLog(log string) error {
+	r, err := os.Open(log)
+	if err != nil {
+		return fmt.Errorf("failed to open log %q: %v", log, err)
+	}
+	defer r.Close()
+	tmpLog := log + tmpSuffix
+	f, err := os.OpenFile(tmpLog, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to create temporary log %q: %v", tmpLog, err)
+	}
+	defer func() {
+		// Best effort cleanup of tmpLog.
+		os.Remove(tmpLog)
+	}()
+	defer f.Close()
+	w := gzip.NewWriter(f)
+	defer w.Close()
+	if _, err := io.Copy(w, r); err != nil {
+		return fmt.Errorf("failed to compress %q to %q: %v", log, tmpLog, err)
+	}
+	compressedLog := log + compressSuffix
+	if err := os.Rename(tmpLog, compressedLog); err != nil {
+		return fmt.Errorf("failed to rename %q to %q: %v", tmpLog, compressedLog, err)
+	}
+	// Remove old log file.
+	if err := os.Remove(log); err != nil {
+		return fmt.Errorf("failed to remove log %q after compress: %v", log, err)
+	}
+	return nil
+}
+
+// rotateLatestLog rotates latest log without compression, so that container can still write
+// and fluentd can finish reading.
+func (c *containerLogManager) rotateLatestLog(id, log string) error {
+	timestamp := c.clock.Now().Format(timestampFormat)
+	rotated := fmt.Sprintf("%s.%s", log, timestamp)
+	if err := os.Rename(log, rotated); err != nil {
+		return fmt.Errorf("failed to rotate log %q to %q: %v", log, rotated, err)
+	}
+	if err := c.runtimeService.ReopenContainerLog(id); err != nil {
+		// Rename the rotated log back, so that we can try rotating it again
+		// next round.
+		// If kubelet gets restarted at this point, we'll lose original log.
+		if renameErr := os.Rename(rotated, log); renameErr != nil {
+			// This shouldn't happen.
+			// Report an error if this happens, because we will lose original
+			// log.
+			glog.Errorf("Failed to rename rotated log %q back to %q: %v, reopen container log error: %v", rotated, log, renameErr, err)
+		}
+		return fmt.Errorf("failed to reopen container log %q: %v", id, err)
+	}
+	if c.policy.MaxFiles > 1 {
+		return nil
+	}
+	// MaxFiles == 1, remove latest rotated log file.
+	// NOTE(random-liu): MaxFiles == 1 is not recommended. Fluentd may not finish
```

Comment:
Do we need to support MaxFiles == 1?  Is there any reason we can't require MaxFiles > 1?

Reply:
I'm not sure whether there will be people want it.
Docker does support it. 
`The maximum number of log files that can be present. If rolling the logs creates excess files, the oldest file is removed. Only effective when max-size is also set. A positive integer. Defaults to 1.`

Reply:
I think if allowing MaxFiles == 1 is requested, we can add it easily.  I would rather do that than implicitly allowing something that doesn't work well with log aggregators.  Docker defaults to 1, so it makes sense for them to allow it.  If we are recommending something different (by using a different default), I don't think we need to allow 1 as an option.

Reply:
OK. Then let's enforce `>1`

Reply:
Done